src/event/HandlerList.php

```diff
@@ -128,11 +130,26 @@ public function getListenerList() : array{
 		}
 
 		$listenersByPriority = [];
+		$asyncListenersByPriority = [];
 		foreach($handlerLists as $currentList){
 			foreach($currentList->handlerSlots as $priority => $listeners){
-				$listenersByPriority[$priority] = array_merge($listenersByPriority[$priority] ?? [], $listeners);
+				$syncListeners = array_filter($listeners, static function(RegisteredListener $listener) : bool{ return !($listener instanceof RegisteredAsyncListener); });
+				$listenersByPriority[$priority] = array_merge($listenersByPriority[$priority] ?? [], $syncListeners);
+
+				$asyncListeners = array_filter($listeners, static function(RegisteredListener $listener) : bool{ return $listener instanceof RegisteredAsyncListener; });
+				$asyncListenersByPriority[$priority] = array_merge($asyncListenersByPriority[$priority] ?? [], $asyncListeners);
 			}
 		}
+		foreach($asyncListenersByPriority as $priority => $asyncListeners){
+			usort($asyncListeners, static function(RegisteredAsyncListener $a, RegisteredAsyncListener $b) : int{
+				if($a->canBeCallConcurrently()){
+					return $b->canBeCallConcurrently() ? 0 : -1;
```

Comment:
Is it arbitrary that exclusive handlers are called first and concurrent handlers are called together last? Would like to see the justification in comments.

Reply:
I have no justification for the fact that these are the exclusive handlers after the concurrent handlers because I haven't been able to make up my mind? The only performance requirement is that the concurrent handlers be grouped together and that the exclusive handlers be called either before or after the concurrent ones.